spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -3038,45 +2954,44 @@ private void doSendOffsets(Producer<?, ?> prod, Map<TopicPartition, OffsetAndMet
 		private void processCommits() {
 			this.count += this.acks.size();
 			handleAcks();
-			AckMode ackMode = this.containerProperties.getAckMode();
-			if (!this.isManualImmediateAck) {
-				if (!this.isManualAck) {
-					updatePendingOffsets();
```

Comment:
Where did this code (`updatePendingOffsets`) move to? Do we not need this any longer? I see that you removed the method, but where is this logic now?

Reply:
`updatePendingOffsets()` feel like duplicate in `processAck()`

https://github.com/spring-projects/spring-kafka/blob/5beb8fe607bd131ced632dbc7afa32e11a6ff2dd/spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java#L2024-L2034

Reply:
But `processAck()` is not called from the hierarchy of `processCommits` where `updatePendingOffsets()` was originally called. I see that there are some duplicates, but still wondering whether it changes the logic a little bit?

Reply:
@Wzy19930507 Any updates on this? This is the only pending item that needs an explanation before merging the PR.

Reply:
Sorry for late.  `handleAcks` has three scenarios and include `updatePendingOffsets` logic

1. `!Thread.currentThread().equals(this.consumerThread)` --> keep `isManualImmediateAck` thread safe.
2. `isManualImmediateAck` --> call `ackImmediate`
3. not `isManualImmediateAck` --> call `addOffset`

`updatePendingOffsets` logic would be:
```
			if (!this.isManualImmediateAck) {
				if (!this.isManualAck) {
					updatePendingOffsets();
				}
```
Look like same as scenarios three.

Reply:
Yes, Yes, totally right, your explanation is clearer!